### PR TITLE
ci: Enable CodeQL scanning

### PR DESCRIPTION
**Pull request by patzielinski**
This PR has CodeQL run as CI on pull requests.

### CODE REVIEW

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    runs-on: ${{ 'ubuntu-latest' }}
+    timeout-minutes: ${{ 360 }}
+    permissions:
+      # required for all workflows
+      security-events: write
+
+      # required to fetch internal or private CodeQL packs
+      packages: read
+
+      # only required for workflows in private repositories
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:39]: we can take these lines out I think?

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        # Use `c-cpp` to analyze code written in C, C++ or both
+        # Use 'java-kotlin' to analyze code written in Java, Kotlin or both
+        # Use 'javascript-typescript' to analyze code written in JavaScript, TypeScript or both
+        # To learn more about changing the languages that are analyzed or customizing the build mode for your analysis,
+        # see https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/customizing-your-advanced-setup-for-code-scanning.
+        # If you are analyzing a compiled language, you can modify the 'build-mode' for that language to customize how
+        # your codebase is analyzed, see https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/codeql-code-scanning-for-compiled-languages
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v4
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:59]: hashes, thanks

File: .github/workflows/codeql.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v3
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:63]: here too

File: .github/workflows/codeql.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          'languages you are analyzing, replace this with the commands to build' \
+          'your code, for example:'
+        echo '  make bootstrap'
+        echo '  make release'
+        exit 1
+
+    - name: Perform CodeQL Analysis
+      uses: github/codeql-action/analyze@v3
+      with:
+        category: "/language:${{matrix.language}}"
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:93]: single trailing newline

File: .github/workflows/codeql.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      run: |
+        echo 'If you are using a "manual" build mode for one or more of the' \
+          'languages you are analyzing, replace this with the commands to build' \
+          'your code, for example:'
+        echo '  make bootstrap'
+        echo '  make release'
+        exit 1
+
+    - name: Perform CodeQL Analysis
+      uses: github/codeql-action/analyze@v3
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:91]: hash

File: .github/workflows/codeql.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        # For more details on CodeQL's query packs, refer to: https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
+        # queries: security-extended,security-and-quality
+
+    # If the analyze step fails for one of the languages you are analyzing with
+    # "We were unable to automatically build your code", modify the matrix above
+    # to set the build mode to "manual" for that language. Then modify this step
+    # to build your code.
+    # ℹ️ Command-line programs to run using the OS shell.
+    # 📚 See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun
+    - if: matrix.build-mode == 'manual'
```
Comment (adityasaky) [inline .github/workflows/codeql.yml:80]: We can probably take all of these out, our build command isn't really special.

**Review by adityasaky — APPROVED** [review]

**Review by neilnaveen — COMMENTED** [review]
Do we want to include scanning on python since our doc tests use python?

**Comment by adityasaky** [comment]
I don't think we need to worry about a test script that's only used on our docs...

**Review by neilnaveen — APPROVED** [review]

**Pull request merged**